mne/tests/test_surface.py

```diff
@@ -236,8 +239,104 @@ def test_voxel_neighbors():
     image = np.zeros((10, 10, 10))
     image[4:7, 4:7, 4:7] = 3
     image[5, 5, 5] = 4
-    volume = voxel_neighbors((5.5, 5.1, 4.9), image, thresh=2)
+    volume = _voxel_neighbors((5.5, 5.1, 4.9), image, thresh=2,
+                              max_peak_dist=1, voxels_max=100)
     true_volume = set([(5, 4, 5), (5, 5, 4), (5, 5, 5), (6, 5, 5),
                        (5, 6, 5), (5, 5, 6), (4, 5, 5)])
     assert volume.difference(true_volume) == set()
     assert true_volume.difference(volume) == set()
+
+
+@requires_nibabel()
+@requires_dipy()
+@pytest.mark.slowtest
+@testing.requires_testing_data
+def test_warp_montage_volume():
+    """Test warping an montage based on intracranial electrode positions."""
+    import nibabel as nib
+    subject_T1 = nib.load(
+        op.join(subjects_dir, 'sample', 'mri', 'T1.mgz'))
+    subject_brain = nib.load(
+        op.join(subjects_dir, 'sample', 'mri', 'brain.mgz'))
+    template_brain = nib.load(
+        op.join(subjects_dir, 'fsaverage', 'mri', 'brain.mgz'))
+    reg_affine, sdr_morph = compute_volume_registration(
+        subject_brain, template_brain, zooms=5,
+        niter=dict(affine=[10, 10, 5], sdr=[3, 3, 3]),
+        pipeline=('affine', 'sdr'))
+    # make fake image with three coordinates
+    CT_data = np.zeros(subject_brain.shape)
+    # make electrode contact hyperintensities
+    CT_data[45:47, 39:41, 49:50] = 500  # surround high intensity
+    CT_data[46, 40, 49] = 1000  # center even higher intensity
+    CT_data[47:49, 39:40, 49:50] = 500
+    CT_data[48, 39, 50] = 1000
+    CT_data[50:52, 38:40, 50:51] = 500
+    CT_data[50, 39, 50] = 1000
+    CT = nib.Nifti1Image(CT_data, subject_T1.affine)
+    ch_coords = np.array([[-8.7040273, 17.99938754, 10.29604017],
+                          [-14.03007764, 19.69978401, 12.07236939],
+                          [-21.1130506, 21.98310911, 13.25658887]])
+    ch_pos = dict(zip(['1', '2', '3'], ch_coords / 1000))  # mm -> m
+    montage = make_dig_montage(ch_pos, coord_frame='mri')
+    montage_warped, image_from, image_to = warp_montage_volume(
+        montage, CT, reg_affine, sdr_morph, 'sample',
+        subjects_dir=subjects_dir, thresh=0.99)
+    # checked with nilearn plot from `tut-ieeg-localize`
+    # check montage in surface RAS
+    np.testing.assert_almost_equal(
+        montage_warped.dig[0]['r'], np.array([-0.31, 0.29, -0.35]), decimal=1)
```

Comment:
`decimal=1` means `0.1` i.e., 10 cm (units here are in meters). Surely this can be better? Also you don't need to explicitly `np.array` this and we always `from numpy.testing import assert_almost_equal` so I would do:
```
    assert_almost_equal(montage_warped.dig[0]['r'], [-0.31, 0.29, -0.35], decimal=3)
```
but this fails
```
mne/tests/test_surface.py:287: in test_warp_montage_volume
    np.testing.assert_almost_equal(
E   AssertionError: 
E   Arrays are not almost equal to 3 decimals
E   
E   Mismatched elements: 3 / 3 (100%)
E   Max absolute difference: 0.062
E   Max relative difference: 0.2137931
E    x: array([-0.313,  0.352, -0.325])
E    y: array([-0.31,  0.29, -0.35])
```
This actually seems quite far off (5cm!)

Reply:
So I can make the tests a lot more precise but it depends a lot on the SDR and how many iterations. I thought it was rather frustrating how precise the arbitrary numbers were in the SDR morph for ranges of numbers so I tried to make this test a bit looser so that it wouldn't fail in the future if things were changed slightly since it's really not that precise unless you use many more iterations. i.e. at this level of test with the roughness it is actually off by 5 cm from the high-precision truth so I think it makes sense to make tests that are appropriately loose. Happy to change to be more precise if that's not convincing to you.

Reply:
5 cm is really far. I would expect even an affine transformation alone (and probably even a rigid affine!) to be off by less for any given point. Hence this makes me worry a bit that something is systematically wrong with the code or the test...